Update requirements and limitations of config when property (#2257)
replicatedhq/replicated-docs: docs/reference/custom-resource-config.mdx

### `when`

The `when` property denotes groups that are displayed on the admin console **Config** page only when a condition evaluates to true. When the condition evaluates to false, the group is not displayed.

<PropertyWhen/>

:::note
`when` is a property of both groups and items. See [Item Properties > `when`](/reference/custom-resource-config#when-item) below.
:::

#### Requirements and Limitations

<WhenRequirements/>

#### Example

> The following example shows a group of items that is displayed on the **Config** page only when the user enables the `http_enabled` configuration field. This example uses the KOTS [ConfigOptionEquals](/reference/template-functions-config-context#configoptionequals) template function to evaluate the value of the `http_enabled` configuration field.

Reviewer: I like how the items `when` example says specifically what will show as a result of what condition. NBD if you use my wording.
```suggestion
In the following example, the `example_group_2` group of items will be displayed on the **Config** page only when the user enables the `http_enabled` configuration field. This example uses the KOTS [ConfigOptionEquals](/reference/template-functions-config-context#configoptionequals) template function to evaluate the value of the `http_enabled` configuration field.
```

<PropertyWhen/>

For examples of configuring the `when` property, see [Using Conditional Statements in Configuration Fields](/vendor/config-screen-conditional).

#### Requirements and Limitations

`when` has the following requirements and limitations:
* The `when` property accepts the following types of values:
   * Booleans
   * Strings that match "true", "True", "false", or "False"

>    [KOTS template functions](/reference/template-functions-about) can be used to render these supported value types.

Author: ^ it looks like both "true" as a string and true as a boolean worked when I tested it. Integers didn't work

This linter rule in the vendor portal it seems to confirm: https://github.com/replicatedhq/vandoor/blob/b6116908862b263f4260e96e3186ee0dd117ae30/graphql-api/src/policies/rego/kots-spec-default.rego#L669

Reviewer 2: this aligns with what i've seen too

Reviewer 2: found this section of the code for how we parse the values https://github.com/replicatedhq/kotskinds/blob/main/multitype/quotedbool.go#L21-L22. seems like there's a variety of things that would work. i probably need to test some of these first to confirm, but i wonder if a quoted integer like `"0"` works based on this?

Author: A interesting, I can test that too

Author: Looks like the linter is throwing an error for "0" or "1"

<img width="867" alt="Screenshot 2024-04-15 at 9 11 37 AM" src="https://github.com/replicatedhq/replicated-docs/assets/11523028/a64b2fc8-8582-4758-b685-60ae27d55a7d">

* Values compared in the conditional statement of the `when` property must match exactly without quotes for the `when` property to evaluate to true

> * `when` can be used only as a property of a group or item. `when` is _not_ an item type and so cannot be used as an option under `select_one`

Author: > so cannot be used as an option under `select_one`

this was here before so I imagine this is a common issue that people have run into

<PropertyWhen/>

For examples of configuring the `when` property, see [Using Conditional Statements in Configuration Fields](/vendor/config-screen-conditional).

#### Requirements and Limitations

`when` has the following requirements and limitations:
* The `when` property accepts the following types of values:
   * Booleans
   * Strings that match "true", "True", "false", or "False"

   [KOTS template functions](/reference/template-functions-about) can be used to render these supported value types.

> * Values compared in the conditional statement of the `when` property must match exactly without quotes for the `when` property to evaluate to true

Reviewer 2: i see this was moved from the partial [here](https://github.com/replicatedhq/replicated-docs/pull/2257/files#diff-639ebb6d1ebbbd99eead40d225e5bda26e5ea17ac273181b68a9e183db84096cL1), but i'm not sure i'm understanding what this call out is supposed to be communicating exactly

Author: Yeah this was here before, and to me it sounded like it was trying to say that "even slight variations will render false so be sure that they match exactly". Not sure if this is maybe obvious enough that it could be removed?

* `when` can be used only as a property of a group or item

> * `when` is _not_ an item type and cannot be used as a standalone item, including as an option under `select_one` items

Author: >including as an option under `select_one` items

this was called out before, so I'm assuming it's a common point of confusion. I updated the description around it to more generically explain that `when` can't be used as an item

#### Example

The following example shows a group of items that is displayed on the **Config** page only when the user enables the `http_enabled` configuration field. This example uses the KOTS [ConfigOptionEquals](/reference/template-functions-config-context#configoptionequals) template function to evaluate the value of the `http_enabled` configuration field.

> 

Author: ^ added a description

    - name: example_group_2
      title: Second Group

>       description: This is descriptive help text.

Reviewer: Nbd to fix here since it might involve a new screenshot, but it would help if these descriptions weren't here or were different for each group. I actually thought this sample was wrong because I went looking for that description and saw it in the screenshot, but it was from the third group and not the second group.

Author: Made a story for this https://app.shortcut.com/replicated/story/102871/update-descriptions-in-the-config-custom-resource-examples

>       # The group is displayed only when the `http_enabled` field is selected 

Reviewer: ```suggestion
      # This group is displayed only when the `http_enabled` field is selected 
```

    - name: example_group_2
      title: Second Group
      description: This is descriptive help text.

>       # The group is not displayed if when is false

Author: ^ strong double negative here

      - name: password_text
        title: Password
        type: password

>         value: '{{repl RandomString 10}}'     

Reviewer: NBD but this line is on this page a few times, but the others still have double quotes

      - name: password_text
        title: Password
        type: password
        value: '{{repl RandomString 10}}'     
```

![Only the first and third groups appear on the config screen](/images/config-screen-group-when-false.png)
[View a larger version of this image](/images/config-screen-group-when-false.png)

For additional examples of configuring the `when` property, see [Using Conditional Statements in Configuration Fields](/vendor/config-screen-conditional).

> 

Author: ^ added a link to the large set of conditional examples

  </tr>    
</table>

### `when` {#when-item}

<table>
  <tr>
    <th>Description</th>

>     <td><p>The <code>when</code> property denotes configuration groups or items that are displayed on the admin console <strong>Config</strong> page only when a condition evaluates to true. <code>when</code> is a property of both groups and items. For more information about the <code>when</code> property, including requirements and limitations, see <a href="/reference/custom-resource-config#when">Group Properties > `when`</a> above.</p></td>

Author: ^ changed this description to explain that it applies to groups and items, and points the user to the Group > when section for more info
